lib/Sema/PreCheckExpr.cpp

```diff
@@ -586,10 +586,22 @@ Expr *TypeChecker::resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE,
     }
 
     auto emitBasicError = [&] {
-      Context.Diags
-          .diagnose(Loc, diag::cannot_find_in_scope, Name,
-                    Name.isOperator())
-          .highlight(UDRE->getSourceRange());
+      SmallString<32> Buffer;
+      StringRef NameString = Name.getString(buffer);
+      // `self` gets diagnosed with a different error when it can't be found.
+      bool diagSelfCannotBeFound = (NameString.compare(StringRef("self")) == 0);
```

Comment:
We store commonly used identifiers on the ASTContext, and there's a convenience method `isSimpleName` for checking that the name isn't compound and matches a given base name, so this could be written as:

```suggestion
      // `self` gets diagnosed with a different error when it can't be found.
      bool diagSelfCannotBeFound = Name.isSimpleName(Context.Id_self);
```